alacritty_terminal/src/grid/resize.rs

```diff
@@ -0,0 +1,367 @@
+        for elem in self {
+            elem.wrapping_shift_left(num_cols, line, delta);
+        }
+    }
+
+    fn shift_right(&mut self, line: Line, delta: usize) {
+        for elem in self {
+            elem.shift_right(line, delta);
+        }
+    }
+
+    fn shrink_cols(&mut self, num_cols: Column) {
+        for elem in self {
+            elem.shrink_cols(num_cols);
+        }
+    }
+}
+
+impl<T: GridCell + Default + PartialEq + Copy> Grid<T> {
+    /// Resize the grid's width and/or height.
+    pub fn resize(
+        &mut self,
+        reflow: bool,
+        lines: Line,
+        cols: Column,
+        resize: &mut [&mut dyn Resize],
+    ) {
+        match self.lines.cmp(&lines) {
+            Ordering::Less => self.grow_lines(lines, resize),
+            Ordering::Greater => self.shrink_lines(lines, resize),
+            Ordering::Equal => (),
+        }
+
+        match self.cols.cmp(&cols) {
+            Ordering::Less => self.grow_cols(reflow, cols, resize),
+            Ordering::Greater => self.shrink_cols(reflow, cols, resize),
+            Ordering::Equal => (),
+        }
+    }
+
+    /// Add lines to the visible area
+    ///
+    /// Alacritty keeps the cursor at the bottom of the terminal as long as there
+    /// is scrollback available. Once scrollback is exhausted, new lines are
+    /// simply added to the bottom of the screen.
+    fn grow_lines(&mut self, new_line_count: Line, resize: &mut [&mut dyn Resize]) {
+        let lines_added = new_line_count - self.lines;
+
+        // Need to resize before updating buffer
+        self.raw.grow_visible_lines(new_line_count, Row::new(self.cols, T::default()));
+        self.lines = new_line_count;
+
+        let history_size = self.history_size();
+        let from_history = min(history_size, lines_added.0);
+
+        // Move existing lines up for every line that couldn't be pulled from history
+        let range = Line(0)..Line(std::usize::MAX);
+        if from_history != lines_added.0 {
+            let delta = lines_added - from_history;
+            self.scroll_up(&(Line(0)..new_line_count), delta, T::default());
+            resize.rotate(self.lines, self.cols, &range, delta.0 as isize);
+        }
+
+        // Move cursor down for every line pulled from history
+        self.cursor_save.point.line += from_history;
+        self.cursor.point.line += from_history;
+
+        self.display_offset = self.display_offset.saturating_sub(*lines_added);
+        self.decrease_scroll_limit(*lines_added);
+    }
+
+    /// Remove lines from the visible area
+    ///
+    /// The behavior in Terminal.app and iTerm.app is to keep the cursor at the
+    /// bottom of the screen. This is achieved by pushing history "out the top"
+    /// of the terminal window.
+    ///
+    /// Alacritty takes the same approach.
+    fn shrink_lines(&mut self, target: Line, resize: &mut [&mut dyn Resize]) {
+        let shrinkage = self.lines.0 - target.0;
+        self.lines = target;
```

Comment:
This needs to be moved to the end, otherwise `self.scroll_up` will be called with the new lines already set, increasing the scrollback history beyond the limits causing a subtract with overflow error.